src/optimizer/rewriter.cpp

```diff
@@ -0,0 +1,150 @@
+//===----------------------------------------------------------------------===//
+//
+//                         Peloton
+//
+// rewriter.cpp
+//
+// Identification: src/optimizer/rewriter.cpp
+//
+// Copyright (c) 2015-2018, Carnegie Mellon University Database Group
+//
+//===----------------------------------------------------------------------===//
+
+#include <memory>
+
+#include "optimizer/optimizer.h"
+#include "optimizer/rewriter.h"
+#include "common/exception.h"
+
+#include "optimizer/cost_model/trivial_cost_model.h"
+#include "optimizer/operator_visitor.h"
+#include "optimizer/optimize_context.h"
+#include "optimizer/optimizer_task_pool.h"
+#include "optimizer/rule.h"
+#include "optimizer/rule_impls.h"
+#include "optimizer/optimizer_metadata.h"
+#include "optimizer/absexpr_expression.h"
+#include "expression/abstract_expression.h"
+#include "expression/constant_value_expression.h"
+
+using std::vector;
+using std::unordered_map;
+using std::shared_ptr;
+using std::unique_ptr;
+using std::move;
+using std::pair;
+using std::make_shared;
+
+namespace peloton {
+namespace optimizer {
+
+using OptimizerMetadataTemplate = OptimizerMetadata<AbsExpr_Container,ExpressionType,AbsExpr_Expression>;
+
+using OptimizeContextTemplate = OptimizeContext<AbsExpr_Container,ExpressionType,AbsExpr_Expression>;
+
+using OptimizerTaskStackTemplate = OptimizerTaskStack<AbsExpr_Container,ExpressionType,AbsExpr_Expression>;
+
+using TopDownRewriteTemplate = TopDownRewrite<AbsExpr_Container,ExpressionType,AbsExpr_Expression>;
+
+using BottomUpRewriteTemplate = BottomUpRewrite<AbsExpr_Container,ExpressionType,AbsExpr_Expression>;
+
+using GroupExpressionTemplate = GroupExpression<AbsExpr_Container,ExpressionType,AbsExpr_Expression>;
+
+using GroupTemplate = Group<AbsExpr_Container,ExpressionType,AbsExpr_Expression>;
+
```

Comment:
Would be nice to see more documentation of the functions in Rewriter as well explanations for the templates you use. Maybe a high level explanation of what the rewriter does at the top of the rewriter.h file would be helpful?